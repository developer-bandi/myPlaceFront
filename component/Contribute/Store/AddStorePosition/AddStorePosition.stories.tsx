import React, { useEffect, useRef } from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import AddStorePosition from "./AddStorePosition";
import { action } from "@storybook/addon-actions";
import { useIsTabletOrMobile } from "../../../../lib/customHook/mediaQuery";

export default {
  title: "Contribute/Store/AddStorePosition",
  component: AddStorePosition,
} as ComponentMeta<typeof AddStorePosition>;

const Template: ComponentStory<typeof AddStorePosition> = (arg) => {
  const isTabletOrMobile = useIsTabletOrMobile();
  const ref = useRef<HTMLDivElement>(null);
  useEffect(() => {
    setTimeout(() => {
      if (ref.current !== null) {
        const div = document.createElement("div");
        div.style.backgroundColor = "black";
        div.style.width = "100%";
        div.style.height = "100%";
        ref.current.appendChild(div);
      }
    }, 2000);
  });
  return (
    <AddStorePosition
      {...arg}
      isTabletOrMobile={isTabletOrMobile}
      mapref={ref}
    />
  );
};

export const generalType = Template.bind({});
generalType.args = {
  setAddress: action("setAddress"),
  moveSetpage: action("moveSetpage"),
};
