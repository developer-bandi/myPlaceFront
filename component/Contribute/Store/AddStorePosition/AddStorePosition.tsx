import {RefObject} from "react";
import storePosition from "../../../../lib/styles/storePosition.module.scss";
interface SideMapProps {
  mapref: RefObject<HTMLDivElement>;
  addressInputRef: RefObject<HTMLInputElement>;
  setAddress: () => void;
  moveSetpage: () => void;
  isTabletOrMobile: boolean;
}

const AddStorePosition = ({
  mapref,
  addressInputRef,
  setAddress,
  moveSetpage,
  isTabletOrMobile,
}: SideMapProps) => {
  return (
    <div className={storePosition.mainBlock}>
      <h1 className={storePosition.title}>장소 등록</h1>
      <div
        ref={mapref}
        style={
          isTabletOrMobile
            ? {width: "100%", height: "400px"}
            : {width: "800px", height: "800px"}
        }
        className={storePosition.map}
      ></div>
      <div className={storePosition.addressBlock}>
        <input
          ref={addressInputRef}
          className={storePosition.addressInput}
          placeholder="지도를 클릭하거나 주소입력후 우측 버튼을 누르세요"
        />
        <button
          className={storePosition.addressButton}
          onClick={() => {
            setAddress();
          }}
          data-testid="setAddress"
        >
          주소 설정
        </button>
      </div>
      <div className={storePosition.submitButtonBlock}>
        <button
          className={storePosition.submitButton}
          onClick={() => {
            moveSetpage();
          }}
          data-testid="moveSetpage"
        >
          이 위치로 장소 설정후 다음단계로 이동
        </button>
      </div>
    </div>
  );
};

export default AddStorePosition;
