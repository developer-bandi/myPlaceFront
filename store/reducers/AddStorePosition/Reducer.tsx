import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { HYDRATE } from "next-redux-wrapper";

export interface AddStorePositionState {
  address: string;
  longitude: string;
  latitude: string;
}

const initialState: AddStorePositionState = {
  address: "",
  longitude: "",
  latitude: "",
};

const AddStorePositionSlice = createSlice({
  name: "addStorePosition",
  initialState,

  reducers: {
    setPosition(state, action: PayloadAction<AddStorePositionState>) {
      state.address = action.payload.address;
      state.latitude = action.payload.latitude;
      state.longitude = action.payload.longitude;
    },
  },

  extraReducers: {
    [HYDRATE]: (state, action) => {
      return state;
    },
  },
});

export const { setPosition } = AddStorePositionSlice.actions;

export default AddStorePositionSlice.reducer;
