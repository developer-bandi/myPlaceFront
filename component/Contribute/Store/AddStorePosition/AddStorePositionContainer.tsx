import { useIsTabletOrMobile } from "../../../../hooks/mediaQuery";
import AddStorePosition from "./AddStorePosition";
import useAddStorePosition from "./AddStorePositionHook";

const AddStorePositionContainer = () => {
  const isTabletOrMobile = useIsTabletOrMobile();
  const { mapref, addressInputRef, setAddress, moveSetpage } =
    useAddStorePosition();

  return (
    <AddStorePosition
      mapref={mapref}
      addressInputRef={addressInputRef}
      setAddress={setAddress}
      moveSetpage={moveSetpage}
      isTabletOrMobile={isTabletOrMobile}
    />
  );
};

export default AddStorePositionContainer;
