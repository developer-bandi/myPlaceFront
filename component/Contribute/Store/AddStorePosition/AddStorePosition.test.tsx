import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
} from "@testing-library/react";
import useAddStorePosition from "./AddStorePositionHook";
import {Provider} from "react-redux";
import configureMockStore from "redux-mock-store";
import newReact from "react";
import * as React from "react";
import {useRouter} from "next/router";
import AddStorePosition from "./AddStorePosition";

jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));
const mockRouter = {
  push: jest.fn(),
};
(useRouter as jest.Mock).mockReturnValue(mockRouter);

describe("AddStorePosition Presentational 테스트", () => {
  const setAddressMock = jest.fn();
  const moveSetpageMock = jest.fn();
  const utils = render(
    <AddStorePosition
      mapref={{current: null}}
      addressInputRef={{current: null}}
      setAddress={setAddressMock}
      moveSetpage={moveSetpageMock}
      isTabletOrMobile={false}
    />
  );
  expect(utils.container).toMatchSnapshot();
  fireEvent.click(screen.getByTestId("setAddress"));
  fireEvent.click(screen.getByTestId("moveSetpage"));
  expect(setAddressMock).toBeCalled();
  expect(moveSetpageMock).toBeCalled();
});

describe("AddStorePosition Hook 테스트", () => {
  let infoBlank = true;
  const mockStore = configureMockStore();
  const StorePositionMockStore = mockStore({
    addStorePosition: {
      address: "",
      longitude: "",
      latitude: "",
    },
  });
  const SettedStorePositionMockStore = configureMockStore()({
    addStorePosition: {
      address: "testAdress",
      longitude: "testLongitude",
      latitude: "testLatitude",
    },
  });
  const wrapper = ({children}: {children: React.ReactNode}) => (
    <Provider
      store={infoBlank ? StorePositionMockStore : SettedStorePositionMockStore}
    >
      {children}
    </Provider>
  );
  describe("setAddress 테스트", () => {
    it("정보를 제대로 찾은 경우", () => {
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({current: {value: "test"}});
      const kakao = {
        maps: {
          services: {
            Geocoder: jest.fn(() => {
              return {
                addressSearch: (input: string, callback: Function) => {
                  callback([{x: "testx", y: "testy"}], "ok");
                },
              };
            }),
            Status: {
              OK: "ok",
              ZERO_RESULT: "zero_result",
              ERROR: "error",
            },
          },
          LatLng: jest.fn(() => "test"),
        },
      };
      window.kakao = kakao as any;
      const setPositionMock = jest.fn();
      const setMapMock = jest.fn();
      const setCenterMock = jest.fn();
      const {result} = renderHook(() => useAddStorePosition(), {wrapper});
      act(() => {
        result.current.setMarker({
          setPosition: setPositionMock,
          setMap: setMapMock,
        });
      });
      act(() => {
        result.current.setMakedMap({
          setCenter: setCenterMock,
        });
      });
      act(() => {
        result.current.setAddress();
      });
      expect(StorePositionMockStore.getActions()[0]).toStrictEqual({
        type: "addStorePosition/setPosition",
        payload: {
          address: "test",
          latitude: "testy",
          longitude: "testx",
        },
      });
    });
    it("결과가 없는 경우", () => {
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({current: {value: "test"}});
      const kakao = {
        maps: {
          services: {
            Geocoder: jest.fn(() => {
              return {
                addressSearch: (input: string, callback: Function) => {
                  callback([{x: "testx", y: "testy"}], "zero_result");
                },
              };
            }),
            Status: {
              OK: "ok",
              ZERO_RESULT: "zero_result",
              ERROR: "error",
            },
          },
          LatLng: jest.fn(() => "test"),
        },
      };
      window.kakao = kakao as any;
      window.alert = jest.fn();
      const setPositionMock = jest.fn();
      const setMapMock = jest.fn();
      const setCenterMock = jest.fn();
      const {result} = renderHook(() => useAddStorePosition(), {wrapper});
      act(() => {
        result.current.setMarker({
          setPosition: setPositionMock,
          setMap: setMapMock,
        });
      });
      act(() => {
        result.current.setMakedMap({
          setCenter: setCenterMock,
        });
      });
      act(() => {
        result.current.setAddress();
      });
      expect(window.alert).toBeCalledWith("주소를 정확하게 입력해주세요");
      expect(StorePositionMockStore.getActions()[1]).toStrictEqual({
        type: "addStorePosition/setPosition",
        payload: {
          address: "",
          latitude: "",
          longitude: "",
        },
      });
      expect(result.current.addressInputRef.current).toEqual({value: ""});
      expect(setMapMock.mock.calls[0][0]).toBe(null);
    });
    it("에러 발생", () => {
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({current: {value: "test"}});
      const kakao = {
        maps: {
          services: {
            Geocoder: jest.fn(() => {
              return {
                addressSearch: (input: string, callback: Function) => {
                  callback([{x: "testx", y: "testy"}], "error");
                },
              };
            }),
            Status: {
              OK: "ok",
              ZERO_RESULT: "zero_result",
              ERROR: "error",
            },
          },
          LatLng: jest.fn(() => "test"),
        },
      };
      window.kakao = kakao as any;
      window.alert = jest.fn();
      const {result} = renderHook(() => useAddStorePosition(), {wrapper});
      act(() => {
        result.current.setAddress();
      });
      expect(window.alert).toBeCalledWith("서버에 에러가 발생하였습니다");
    });
  });
  describe("moveSetpage 함수 테스트", () => {
    it("값이 채워져 있는 경우", () => {
      infoBlank = false;
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({current: {value: "test"}});
      const {result} = renderHook(() => useAddStorePosition(), {wrapper});
      act(() => {
        result.current.moveSetpage();
      });
      expect(mockRouter.push).toBeCalledWith("/contribute/addstoreinfo");
    });
    it("값이 비어 있는 경우", () => {
      infoBlank = true;
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({current: {value: "test"}});
      window.alert = jest.fn();
      const {result} = renderHook(() => useAddStorePosition(), {
        wrapper,
      });
      act(() => {
        result.current.moveSetpage();
      });
      expect(window.alert).toBeCalledWith("위치를 설정해주세요");
    });
  });
});
