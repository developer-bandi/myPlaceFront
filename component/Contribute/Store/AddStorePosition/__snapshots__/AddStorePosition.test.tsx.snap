// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div
    class="mainBlock"
  >
    <h1
      class="title"
    >
      장소 등록
    </h1>
    <div
      class="map"
      style="width: 800px; height: 800px;"
    />
    <div
      class="addressBlock"
    >
      <input
        class="addressInput"
        placeholder="지도를 클릭하거나 주소입력후 우측 버튼을 누르세요"
      />
      <button
        class="addressButton"
        data-testid="setAddress"
      >
        주소 설정
      </button>
    </div>
    <div
      class="submitButtonBlock"
    >
      <button
        class="submitButton"
        data-testid="moveSetpage"
      >
        이 위치로 장소 설정후 다음단계로 이동
      </button>
    </div>
  </div>
</div>
`;
